tempesta_fw/tls.c

```diff
@@ -19,50 +19,250 @@
  * this program; if not, write to the Free Software Foundation, Inc., 59
  * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
+#include "cfg.h"
 #include "connection.h"
+#include "http_msg.h"
 #include "tls.h"
 
+typedef struct {
+	mbedtls_ssl_config	cfg;
+	mbedtls_x509_crt	crt;
+	mbedtls_pk_context	key;
+} TfwTls;
+
+static TfwTls tfw_tls = { 0 };
+
 /**
- * TODO do all crypto and handle TLS FSM here.
- *
  * Decrypted response messages should be directly placed in TDB area
  * to avoid copying.
  */
 static int
 tfw_tls_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
 {
-	int r = TFW_BLOCK;
-	TfwConnection *c = (TfwConnection *)conn;
-	TfwMsg *msg = (TfwMsg *)c->msg;
+	int r;
+	TfwConnection *c = conn;
+	TfwTlsContext *tls = c->tls;
+
+	TFW_DBG("TLS:%p %s\n", tls, __func__);
+
+	ss_skb_queue_tail(&tls->rx_queue, skb);
+
+	r = mbedtls_ssl_handshake(&c->tls->ssl);
+	if (r == MBEDTLS_ERR_SSL_CONN_EOF) {
+		return TFW_PASS; /* more data needed */
+	} else if (r == 0) {
+		struct sk_buff *nskb;
+
+#define MAX_PAYLOAD_SIZE 1460
+
+		nskb = alloc_skb(MAX_TCP_HEADER + MAX_PAYLOAD_SIZE, GFP_ATOMIC);
+		if (unlikely(!nskb))
+			return TFW_BLOCK;
+
+		skb_reserve(nskb, MAX_TCP_HEADER);
+		skb_put(nskb, MAX_PAYLOAD_SIZE);
+
+		r = mbedtls_ssl_read(&tls->ssl, nskb->data, MAX_PAYLOAD_SIZE);
+		if (r == MBEDTLS_ERR_SSL_WANT_READ ||
+		    r == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			kfree_skb(nskb);
+			return TFW_PASS;
+		} else if (r <= 0) {
+			kfree_skb(nskb);
+			return r ? TFW_BLOCK : TFW_PASS;
+		}
 
-	/*
-	 * TODO switch to HTTP FSM, @data and @len must be decrypted here.
-	 * Typically we don't need original skb any more either for server or
-	 * proxy modes, so it has sense to do decryption in-place.
-	 */
-	r = tfw_gfsm_move(&msg->state, TFW_HTTPS_FSM_TODO_ISSUE_81, skb, off);
+		TFW_DBG2("TLS:%p read %d bytes [%.*s]\n", tls, r, r, nskb->data);
 
-	return r;
+		skb_trim(nskb, r);
+
+#undef MAX_PAYLOAD_SIZE
+
+		r = tfw_gfsm_move(&c->msg->state, TFW_HTTPS_FSM_CHUNK, nskb, 0);
+		if (r == TFW_BLOCK)
+			return TFW_BLOCK;
```

Comment:
This is were do we move to state of processing new data chunk which we've just decrypted. HTTP establishes its hook for the TLS state and must call `tfw_http_conn_msg_alloc()` on its own when it's in `TFW_HTTPS_FSM_INIT` - now the state must be initialized by GFSM rather than by `tfw_http_conn_msg_alloc()` (see similar behaviour for `TFW_FRANG_FSM_INIT` when Frang sets HTTP hooks).